python/lsst/ts/dcs/mock_controller.py

```diff
@@ -0,0 +1,249 @@
+
+
+class MockDomeController:
+    """Mock DomeController that talks over TCP/IP.
+
+    Parameters
+    ----------
+    port : int
+        TCP/IP port
+
+    Notes
+    -----
+    To start the server:
+
+        ctrl = MockDomeController(...)
+        await ctrl.start()
+
+    To stop the server:
+
+        await ctrl.stop()
+
+    Known Limitations:
+
+    * Just a framework that needs to be implemented properly
+    """
+
+    def __init__(
+            self,
+            port,
+    ):
+        self.port = port
+        self._server = None
+        self.log = logging.getLogger("MockDomeController")
+        # Dict of command: (has_argument, function).
+        # The function is called with:
+        # * No arguments, if `has_argument` False.
+        # * The argument as a string, if `has_argument` is True.
+        self.dispatch_dict = {
+            "status": (False, self.status),
+            "Dome_command_moveAz": (False, self.dome_command_move_az),
+            "Dome_command_moveEl": (False, self.dome_command_move_el),
+            "Dome_command_stopAz": (False, self.dome_command_stop_az),
+            "Dome_command_stopEl": (False, self.dome_command_stop_el),
+            "quit": (False, self.quit),
+        }
+        # Name of a command to report as failed once, the next time it is seen,
+        # or None if no failures. Used to test CSC handling of failed commands.
+        self.fail_command = None
+
+    async def start(self):
+        """Start the TCP/IP server.
+
+        Set start_task done and start the command loop.
+        """
+        self.log.info("start called")
+        self._server = await asyncio.start_server(
+            self.cmd_loop, host="127.0.0.1", port=self.port
+        )
+        await self._server.serve_forever()
+
+    async def stop(self):
+        """Stop the TCP/IP server.
+        """
+        if self._server is None:
+            return
+
+        server = self._server
+        self._server = None
+        logging.info("closing server")
+        server.close()
+        logging.info("done closing")
+
+    async def cmd_loop(self, reader, writer):
+        """Execute commands and output replies."""
+        self.log.info("cmd_loop begins")
+        writer.write("Enter Command > ".encode())
+        await writer.drain()
+        while True:
+            line = await reader.readline()
+            line = line.decode()
+            if not line:
+                # connection lost; close the writer and exit the loop
+                self.log.info("closing the writer")
+                writer.close()
+                self.log.info("stopping mock controller")
+                await self.stop()
+                return
+            line = line.strip()
+            self.log.info(f"read command: {line!r}")
+            if line:
+                try:
+                    items = line.split()
+                    cmd = items[-1]
+                    if cmd not in self.dispatch_dict:
```

Comment:
Instead of raising an exception (and capturing it out later to log/ignore) I think it would be better to respond with a failure code. This way, whoever sent the command knows it failed and why.

Reply:
Well, I copied that from the mock_controller code in ts_ATDome ;-) However, I agree that returning a failure code would be better. I'll need to discuss this with German and EIE to see how we solve this. Once we decide how I will open a new ticket to solve this and the other Exception as well.